realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -75,6 +84,82 @@ static RealmCacheType valueOf(Class<? extends BaseRealm> clazz) {
         }
     }
 
+    private static class CreateRealmRunnable<T extends BaseRealm> implements Runnable {
+        private RealmConfiguration configuration;
+        private RealmInstanceCallback<T> callback;
+        private Class<T> realmClass;
+        private CountDownLatch fgRealmCreatedLatch = new CountDownLatch(1);
+        private RealmNotifier notifier;
+
+        CreateRealmRunnable(RealmNotifier notifier, RealmConfiguration configuration,
+                            RealmInstanceCallback<T> callback, Class<T> realmClass) {
+            this.configuration = configuration;
+            this.realmClass = realmClass;
+            this.callback = callback;
+            this.notifier = notifier;
+        }
+
+        @Override
+        public void run() {
+            T instance = null;
+            final AtomicBoolean taskCanceled = new AtomicBoolean(false);
+            try {
+                instance = createRealmOrGetFromCache(configuration, realmClass);
+                boolean results = notifier.post(new Runnable() {
+                    @Override
+                    public void run() {
+                        // If the RealmAsyncTask.cancel() is called before, we just return without creating the Realm
+                        // instance on the caller thread.
+                        // Thread.isInterrupted() cannot be used for checking here since CountDownLatch.await() will
+                        // will clear interrupted status.
+                        if (taskCanceled.get()) {
+                            return;
+                        }
+                        T instanceToReturn = null;
+                        try {
+                            instanceToReturn = createRealmOrGetFromCache(configuration, realmClass);
+                        }catch (RuntimeException e) {
+                            callback.onError(e);
+                        } finally {
+                            fgRealmCreatedLatch.countDown();
+                        }
+                        if (instanceToReturn != null) {
+                            callback.onSuccess(instanceToReturn);
+                        }
+                    }
+                });
+                if (!results) {
+                    fgRealmCreatedLatch.countDown();
+                }
+                // There is a small chance that the posted runnable cannot be executed because of the thread terminated
```

Comment:
It can, if other thread call `thatThread.interrupt()` or `asyncTask.cancel()`
This is majorly to protect below case:
```
Event QUEUE
1 some other event
2 getInstance in foreground
```

If 1 some other event clean the event queue and quits the looper, the 2 will never be executed then the async task hangs forever.